lib/buildkite/test_collector/session.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+module Buildkite::TestCollector
+  class Session
+    UPLOAD_THREAD_TIMEOUT = 60
+    UPLOAD_SESSION_TIMEOUT = 60
+
+    def initialize
+      @send_queue_ids = []
+      @upload_threads = []
+    end
+
+    def add_example_to_send_queue(id)
+      @send_queue_ids << id
+
+      if @send_queue_ids.size >= Buildkite::TestCollector.batch_size
+        send_ids = @send_queue_ids.shift(Buildkite::TestCollector.batch_size)
+        upload_data(send_ids)
+      end
+    end
+
+    def send_remaining_data
+      return if @send_queue_ids.empty?
+
+      upload_data(@send_queue_ids)
+    end
+
+    def close
+      # There are two thread joins here, because the inner join will wait up to
+      # UPLOAD_THREAD_TIMEOUT seconds PER thread that is uploading data, i.e.
+      # n_threads x UPLOAD_THREAD_TIMEOUT latency if Buildkite happens to be
+      # down. By wrapping that in an outer thread join with the
+      # UPLOAD_SESSION_TIMEOUT, we ensure that we only wait a max of
+      # UPLOAD_SESSION_TIMEOUT seconds before the session exits.
+      Thread.new do
+        @upload_threads.each { |t| t.join(UPLOAD_THREAD_TIMEOUT) }
```

Comment:
My understanding of `join` is that any exceptions from the thread it is called on are now processed in this thread. Is this ok? If so then this approach is the easiest to reason with.

Reply:
Mmm yeah, I guess we could wrap this entire block in a begin rescue block and catch StandardError, so whatever exceptions that could get thrown in the thread don't make the rspec job blow up? 

Maybe while we're just testing this on buildkite/buildkite we don't do that so we can see what errors we're getting and deal with them appropriately, but before this gem is released for customer use we put in a protection like that so we don't break any customer builds?

Reply:
Based on my reading of the docs for the `join` method, exceptions are not raised if [::abort_on_exception](https://github.com/buildkite/test-collector-ruby/pull/174/Thread.html#method-c-abort_on_exception) is set to false.  And it seems the default value for `abort_on_exception` is `false`.

If you need to catch StandardError, you could consider making the `upload` method handle it itself inside the thread.

As a further note - there may also be other rescues for StandardError further up the call stack as well due to the strategy of not breaking builds if there is an exceptional error in this Ruby Collector.

Reply:
Report on exception defaults to true: https://ruby-doc.org/core-2.5.0/Thread.html#method-c-report_on_exception

Reply:
@gchan and I caught up about this 

[abort_on_exception](https://ruby-doc.org/3.0.5/Thread.html#method-c-abort_on_exception) controls whether errors in threads are raised _immediately_ to the main thread or not. This defaults to false.

When `join` is called on a thread that has thrown an exception, that [exception is then re-raised on the main thread](https://ruby-doc.org/3.0.5/Thread.html#class-Thread-label-Exception+handling)

While we are still testing and only using this new version of the gem internally, it's better to have more visibility on the exceptions so we can deal with any errors. Before public release of this gem, wrap this bit of code, and any other risky bits of code in the [safe method](https://github.com/buildkite/test-collector-ruby/blob/4f2ee8dae18274ecceb1aa3ffadedc1d60591ab4/lib/buildkite/test_collector.rb#L109) or similar